### ISSUE TITLE
Actual podcast times differ depending on media player

### ISSUE THREAD

**Issue by jpgill86** [critrolesync/critrolesync.github.io#5]
The same exact podcast audio file played by different media players or in different browsers can have different timestamps.

Here are some numbers I collected for the start of the intermission in C1E73:

![image](https://user-images.githubusercontent.com/241234/86314279-46f40400-bbf5-11ea-8f77-07132ffd6eb1.png)

Podcast links used for this test:
- [CritRole.com](http://traffic.libsyn.com/critrole/073_CR_podcast_44_128.mp3)
- [Podbean](https://criticalrole.podbean.com/e/episode-73-the-coming-storm/)
- [Google Play](https://play.google.com/music/m/Dhqgijjunyxcdr5bkghi3ihcz4u?t=Vox_Machina_Ep_73_-_The_Coming_Storm-Critical_Role)
- [Apple Podcasts](https://podcasts.apple.com/us/podcast/vox-machina-ep-73-the-coming-storm/id1243705452?i=1000394889114)
- [Podcast Republic](https://www.podcastrepublic.net/podcast/1243705452)

Audio files can be encoded with a "variable bitrate" (VBR), which, according to [this blog post](https://medium.com/@mattbasta/vbr-is-it-actually-as-bad-as-they-say-9ee55ff477de), can make it difficult for media players to accurately jump to specific timecodes. I thought this must be the problem, and that Firefox+VLC use one algorithm for handling VBR, and Chrome+PR use another. However, when I inspected the MP3 file for this episode with some advanced tools (exiftool and vbrfix), I found that it was not encoded with a variable bitrate.

So, I really don't know what the true root of the problem is yet. Worse, I don't know which set of timestamps to prefer. Perhaps the right attitude to take here would be "either is good enough", but I have tasted the deliciousness of precise-to-the-word media switching and don't like the idea of settling for less. 😭 Unfortunately, our users' podcast media player is outside of our control, and until I understand the root cause of the variations there's little I can do.

I am considering picking one (Chrome-like timestamps or Firefox-like timestamps) and sticking with it. Before picking, I'd like to review the timestamps already collected to see which category they fall into, as well as check to see if these discrepancies are consistent across campaigns 1 and 2. If we're lucky, these discrepancies may happen only in a minority of episodes.

**Comment by jpgill86** [comment]
Thanks to @theDragonTurtle for helping bring this issue to light.

**Comment by jpgill86** [comment]
My understanding of this issue has improved over the last week.

**TL;DR: Some podcast media players will play the same audio file slightly slower than other media players, causing the entire episode duration to be increased by 0.23% and for all timestamps to occur later by the same factor. To get around this, I've added an option to the app that lets users specify their type of podcast player so that times can be converted consistently.**

The long version:

As an example, here's a link to the podcast for C2E22, at the 3-hour mark: https://overcast.fm/+QzWYHkhYw/3:00:00

The problem is that this points to different moments in the episode if opened in Chrome vs Firefox.

If I open that link in Firefox, I hear "... grabbed it and jammed it into his chest".

If I open it in Chrome, I hear "... as you glance around you feel tension on your body".

These moments are actually separated by 25 seconds.

This issue isn't specific to this particular podcast player (Overcast). If I obtain the MP3 files from other podcast sources, I see the same discrepancies. And this problem goes beyond Firefox vs Chrome. We (again, big thanks to @theDragonTurtle) have seen differences in timestamp resolution when the same podcast app is used on different operating systems (Android vs iOS).

I'm not an audio engineer, but I've tried looking closely at the MP3 files and have developed some hunches. It seems that the MP3s' metadata report a bitrate of 128 kbps, but I think maybe Firefox (and VLC, by the way) ignores this metadata and decodes it at 127.7 kbps.

Furthermore, Firefox can change its estimate of the episode duration. If I try to stream the podcast directly from the host (Podbean; I'm not seeing this particular effect when streaming on Overcast), I can see that Chrome will report a fixed episode duration immediately which never changes. Given the file size, I've worked out that this duration corresponds to 128 kbps. In contrast, Firefox will start with that same episode duration but then change it if you advance the playback position to the end. The new duration corresponds to an average bitrate of about 127.7 kbps. This duration is 128/127.7 = 0.23% longer, and all timestamps have to be increased by this same factor to locate the same moment.

These discrepancies apply to episodes C2E65 and earlier. For recent episodes (C2E66 and later), none of this seems to happen, and timestamps are identical in Firefox and Chrome. Apparently the files are formatted differently somehow.

To deal with these discrepancies, [I now request](https://github.com/critrolesync/critrolesync.github.io/issues/1#issuecomment-655285824) that Chrome is used for all timestamp collection. All timestamps stored in [data.json](https://github.com/critrolesync/critrolesync.github.io/blob/master/data.json) will be taken from Chrome-like 128 kbps constant bitrate media players. Converting times for 127.7 kbps average bitrate media players is easy (just increase or decrease all times by the 0.23% factor), so I've added a "Podcast decoder mode" switch to the app for selecting constant bitrate (CBR) or average bitrate (ABR).

**Comment by jpgill86** [comment]
I'd still appreciate an expert audio engineer's explanation for what's going on here, especially relating to why the schism exists between Chrome/Android and Firefox/iOS, but I am going to consider this problem fixed by the addition of the "Podcast decoder mode" switch (fixed by 951695e).

**Comment by jpgill86** [comment]
Sometime around 2020-08-20, episodes C2E20 and later were re-released in M4A format (see [here](https://github.com/critrolesync/critrolesync.github.io/issues/6#issuecomment-677713212) for more details). These new audio files appears to not have the same problem as the original MP3 files. The playback rate seems to be the same in Chrome and Firefox. Although this is a pain in the short term because many timestamps need to be changed, it is probably good in the long term since it eliminates player-dependent discrepancies.
